### ISSUE TITLE
Takes too long to render with large amount of data

### ISSUE THREAD

**Issue by anandchapter247** [valerybugakov/react-selectable-fast#79]
Hi, I have a table, and its cells are selectable. There are 15000 rows and each row has 30 Selectable `cells`.
I am having a strange issue with re-rendering.
It's taking longer to re-render

![captured](https://user-images.githubusercontent.com/8499364/77816238-41af8680-70e7-11ea-8736-47cd07dd84d3.gif)

It's working fine if I don't update the state, As soon as I set state, it is taking long.

![image](https://user-images.githubusercontent.com/8499364/77816267-73285200-70e7-11ea-81ed-5ca96ba003ac.png)

If I remove `setState(updatedState);` it works perfect. How can I fix it?

Btw, `setState` is called in the parent component of the `Selectable` item.

**Comment by anandchapter247** [comment]
@valerybugakov  For reference:

`Versions`
```
react: 16.13.1
react-dom: 16.13.1
react-selectable-fast: 3.3.0
```

`App.js`

```
function App() {
  const [testState, setTestState] = useState([])
  const onSelect = (selected) => {
    if (selected.length) {
      setTestState(selected)
    }
  }
  console.log(testState)
  return (
    <div className="App">
      <Selectable onSelect={onSelect} />
    </div>
  );
}
```

`Selectable.js`

```
import React from 'react'
import Cells from './Cells';
import { SelectableGroup } from 'react-selectable-fast';

const getRows = (number) => {
  const items = [];
  for (let index = 0; index < number; index++) {
    const cells = []
    for (let j = 0; j < 30; j++) {
      cells.push(
        <Cells key={j} />
      )
    }
    items.push(
      <tr key={index} >
        {cells}
      </tr>
    )
  }
  return items
}
const getHead = () => {
  const head = []
  for (let j = 0; j < 30; j++) {
    head.push(
      <th key={j}>Col ${j}</th>
    )
  }
  return head
}
const Selectable = ({ onSelect }) => (
  <SelectableGroup
    allowClickWithoutSelected
    className='custom-row-selector'
    clickClassName='tick'
    resetOnStart={true}
    onSelectionFinish={(selected) => {
      console.time("EEEE")
      console.log("Selection Started")
      onSelect(selected)
      console.timeEnd("EEEE")
    }}
    ignoreList={['.name-col', '.h-col', '.s-col', '.u-col', '.v-col']}
  >
    <table border="1">
      <thead>
        <tr>
          {getHead()}
        </tr>
      </thead>
      <tbody>
        {getRows(1000)}
      </tbody>
    </table>
  </SelectableGroup>
);

export default Selectable;
```

`Cells.js`

```
import React from 'react'
import { createSelectable } from 'react-selectable-fast';

const Cells = ({ selectableRef, isSelected, isSelecting }) => (
  <td ref={selectableRef}>
    <div style={
      {
        color: isSelected || isSelecting ? "#fff" : "#000",
        background: isSelected || isSelecting ? "#000" : "#fff"
      }
    } >Cells {Math.round(Math.random() * 1000)}</div>
  </td>
);

export default createSelectable(Cells);
```

The reason I think is the re-rendering.
Here is the output of `console.timeEnd`

![image](https://user-images.githubusercontent.com/8499364/77820954-21de8980-710c-11ea-84d2-514f707ee664.png)


Please suggest a solution for the issue.

**Comment by valerybugakov** [comment]
@anandchapter247 could you create a minimal example using [codesandbox](https://codesandbox.io/)? I will be happy to help having it on hand

**Comment by anandchapter247** [comment]
Sure! Here it is: https://codesandbox.io/s/unruffled-vaughan-vnox7

**Comment by valerybugakov** [comment]
@anandchapter247 there're a couple of issues: 
1. It's too many items to render and update at once. To handle all these cards interactively you need to use virtualization to avoid rendering all of them. 
2. It's possible to make a little bit faster using some react optimizations like `useCallback, memo` etc. I've added a few of them while I was testing: https://codesandbox.io/s/reverent-brattain-97bdi?file=/src/Selectable.js
3. I've updated the library to avoid some recalculations. It should a bit faster during selection phase now.

Overall it doesn't solve your problem. The library cannot handle those many items at once. I doubt that any would do it efficiently without virtualization. I will close the issue for now, because it's not a library issue. Let me know if you have any further comments.

**Comment by anandchapter247** [comment]
Thanks. I have implemented it with `react-window` and the issue is being fixed. though I am facing another issue, so instead of creating new, I can post it here. 
So, what I want if the user clicks on the selectable item with the control/command key, I need the item to be selected but with previously selected items. But if the user selects an item without the control/command key, I need the selection restart. I used the `allowCtrlClick` and `resetOnStart={false}` which is allowing me to select with the control key pressed but it is not resetting the selected item, to be reset. Is there any way or hack I can achieve this?
